arangod/Aql/RegisterPlan.cpp

```diff
@@ -28,41 +28,101 @@
 #include "Aql/CollectNode.h"
 #include "Aql/ExecutionNode.h"
 #include "Aql/GraphNode.h"
-#include "Aql/IndexNode.h"
 #include "Aql/IResearchViewNode.h"
+#include "Aql/IndexNode.h"
 #include "Aql/ModificationNodes.h"
 #include "Aql/SubqueryEndExecutionNode.h"
 
 #include <velocypack/Builder.h>
-#include <velocypack/Slice.h>
 #include <velocypack/Iterator.h>
+#include <velocypack/Slice.h>
 #include <velocypack/velocypack-aliases.h>
 
 using namespace arangodb;
 using namespace arangodb::aql;
 
 // Requires RegisterPlan to be defined
-VarInfo::VarInfo(int depth, RegisterId registerId)
+VarInfo::VarInfo(unsigned int depth, RegisterId registerId)
     : depth(depth), registerId(registerId) {
-  TRI_ASSERT(registerId < RegisterPlan::MaxRegisterId);
+  TRI_ASSERT(registerId < RegisterPlanT<ExecutionNode>::MaxRegisterId);
+}
+
+template <typename T>
+void RegisterPlanWalkerT<T>::after(T* en) {
+  TRI_ASSERT(en != nullptr);
+
+  if (en->isIncreaseDepth()) {
+    plan->increaseDepth();
+  }
+
+  for (VariableId const& v : en->getOutputVariables()) {
+    if (v != RegisterPlanT<T>::MaxRegisterId) {
+      plan->registerVariable(v);
+    }  else {
+      plan->addRegister();
+    }
+  }
+
+  if (en->getType() == ExecutionNode::SUBQUERY || en->getType() == ExecutionNode::SUBQUERY_END) {
+    plan->addSubqueryNode(en);
+  }
```

Comment:
I noticed you can omit `SUBQUERY_END` here, it'll be ignored anyway later:
```ExecutionNode::planRegisters():
...
  for (auto& s : v->subQueryNodes) {
    if (s->getType() == ExecutionNode::NodeType::SUBQUERY) {
...
    }
  }
```
```suggestion
  if (en->getType() == ExecutionNode::SUBQUERY) {
    plan->addSubqueryNode(en);
  }
```
If we do that, the branch can be from the loop in `planRegisters` as well.